SubSceneクラスと継承クラスの作成

Sharoku-haga opened the pull request:
#30 の作業となります。

主な作業
・SubSceneクラス作成
・上記クラスの継承クラス作成

作業予定時間
2日予定（11/05まで）

ソースチェック
- [x] kakeru4
- [ ] Kawaguchi Soichiro
- [ ] ichikawaryouya0719

Sharoku-haga commented:
現在クラス名で悩んでいます。SubSceneクラスの継承クラスを作成しようとしているのですが、～SubSceneにするか、～Sceneにするか、またはなにもつけずにクラス名にするかの3つで悩んでいます。
例：OpeningSubScene、OpeningScene、Opening
意見お願いします。

kakeru4 commented:
僕はSubとSceneはつけなくていいと思うのでOpeningでいいと思います。

Sharoku-haga commented:
なるほど。何もつけなくてもいいかもしれないね。
その案でいきます。
ありがとう。

taishi-matsumura reviewed (changes requested):
SubもSceneも付けなくて良いですが、その場合は内部クラスにするとかnamespace切るとか工夫して下さい

Commit:
[enh #30] SuBSceneクラスと継承クラス作成

-基底クラスであるSubSceneクラスの作成
-SubSceneクラスの継承クラス8つ作成
-SubSceneクラス関連を名前空間sub_sceneできっている
-コードとコメント修正

Sharoku-haga commented:
SubSceneクラスのコーディングが完了しましたので、レビューお願いします。
チームメンバーはコードレビュー後、問題なければ一番上のコメントにチェックをいれてください。
1人以上チェックがついたら、コメントをまとめて後、kakeru4にマージしてもらいます。
taishi-matsumuraのアドバイスからSubSceneクラスとその継承クラスはnamespace (名：sub_scene)できっております。

taishi-matsumura reviewed (commented):
namespaceで分ける方針、了解です。

.cppは全て下記のように書かれていますが、

``` cpp
sub_scene::Battle::Battle(GameDataManager* pGameDataManager, GameTimer* pGameTimer):
```

.hと同じようにnamespaceで括ってしまった方が見やすいのでは？

``` cpp
namespace sub_scene {

Battle::Battle(GameDataManager* pGameDataManager, GameTimer* pGameTimer):

・
・
}
```

taishi-matsumura reviewed (commented):
これは好みの問題なのですが、、、
自分は(C++では)namespaceの中はインデント下げない方が好きです。
[GoogleのStyleGuide](http://www.textdrop.net/google-styleguide-ja/cppguide.xml#%E5%90%8D%E5%89%8D%E7%A9%BA%E9%96%93%E3%81%AE%E6%9B%B8%E5%BC%8F)も参考にしてください。

taishi-matsumura reviewed (changes requested):
ファイル内の行に対してのコメントに移動しました

~~enum SUBSCENE_IDをSubScene.hで定義するならnamespace sub_sceneの中に入れるべきでしょうし、 namespaceの外に出すならGameSceneとかもっと上位で定義すべきでは？~~

taishi-matsumura reviewed (changes requested):
ファイル内の行に対してのコメントに移動しました

~~ファイルのヘッダコメントですが、~~
~~\- .hの場合「xxxクラスヘッダ」~~
~~\- .cppの場合：「xxxクラス実装」~~

~~とかで良いと思います。(あまり意味の無い形式的なものになってしまいますが、、、)~~
~~参考：http://www.denshin8.jp/den8dev/doxygen.html~~

~~それと、classにコメント付けて下さい。~~
~~これもとりあえず「xxxクラス」というbriefコメントだけで良いです。~~
~~それで意味のわからないクラスについては、そのクラスの責務や使い方等を詳細コメントで書けば良いかと。~~

masami-sugao commented on Twar/Twar/SubScene/SubScene.cpp:
```diff
@@ -0,0 +1,20 @@
+namespace sub_scene
+{
+SubScene::SubScene() :
+m_pGameLib(&GameLib::GetInstance())
```
`GameLib::GetInstance()`はpublic関数で返却値も参照なので、それをメンバ変数として保持しておく必要はないように思います。
使用頻度にもよりますが、使用するところで`GameLib::GetInstance().Hoge()`とすればよいかと。

Sharoku-haga replied:
確かにその通りだと思います。修正しました。

taishi-matsumura commented on Twar/Twar/SubScene/SubScene.h:
```diff
@@ -0,0 +1,48 @@
+	GAME_CLEAR,			//!< ゲームクリア
+	GAME_OVER,			//!< ゲームオーバー
+};
+class SubScene
```
どーーーーーーでもいいですが、
「アドバイスによる修正」よりは「レビュー指摘事項是正」とかの方が良いような、、、

taishi-matsumura replied:
classにコメント付けて下さい。 これもとりあえず「xxxクラス」というbriefコメントだけで良いです。 それで意味のわからないクラスについては、そのクラスの責務や使い方等を詳細コメントで書けば良いかと。

Sharoku-haga replied:
クラスのコメントはSubSceneクラスだけコメントがなかったので追加しました。
確認お願いします。

taishi-matsumura commented on Twar/Twar/SubScene/Battle.cpp:
```diff
@@ -0,0 +1,34 @@
+﻿/**
+* @file Battle.cpp
+* @brief SubSceneの継承クラスであるBattleクラスのcpp
```
ファイルのヘッダコメントですが、
- .hの場合「xxxクラスヘッダ」
- .cppの場合：「xxxクラス実装」

とかで良いと思います。(あまり意味の無い形式的なものになってしまいますが、、、)
参考：http://www.denshin8.jp/den8dev/doxygen.html

Sharoku-haga replied:
コメントを修正しました。
確認お願いします。

masami-sugao commented on Twar/Twar/SubScene/BattleResult.h:
```diff
@@ -0,0 +1,48 @@
+private:
+	SaveDataManager* m_pSaveDataManager;		//!< セーブデータ管理クラス
+	GameDataManager* m_pGameDataManager;		//!< ゲームデータ管理クラス
+	GameTimer*		 m_pGameTimer;				//!< ゲーム時間管理クラス
```
`SubScene`を継承する多くのクラスでメンバ変数に`SaveDataManager`, `GameDataManager`, `GameTimer`のポインタを持っていますが、少し冗長に感じます。
`SubScene`クラスのメンバ変数に持ってしまって、コンストラクタの引数の初期値をNULLにしてしまった方がすっきりしするように思います。

また、別の観点として、各クラスで利用するこの3つのオブジェクトはそれぞれ別のインスタンスなのでしょうか。
クラス名から想像するとゲーム全体で1つのインスタンスになるような気がします。
その場合はシングルトンで作成する方がよいと思います。

Sharoku-haga replied:
SaveDataManager, GameDataManager, GameTimerの3つをSubSceneクラスのメンバ変数に修正しました。
シングルトンですが、レビューを受けてSaveDataManagerはシングルトンを検討しています。シングルトンにする際は別でissuesをきってから行おうと考えています。

taishi-matsumura commented on Twar/Twar/SubScene/SubScene.h:
```diff
@@ -0,0 +1,49 @@
+#define SUB_SCENE_H
+
+/**サブシーンID*/
+enum SUBSCENE_ID
```
`enum SUBSCENE_ID`をSubScene.hで定義するなら`namespace sub_scene`の中に入れるべきでしょうし、namespaceの外に出すなら`GameScene`とかもっと上位で定義すべきでは？

masami-sugao commented on Twar/Twar/SubScene/Battle.h:
```diff
@@ -0,0 +1,45 @@
+	Battle(GameDataManager* pGameDataManager, GameTimer* pGameTimer);
+
+	/**デストラクタ*/
+	virtual~Battle();
```
これでコンパイラが認識できているのであれば問題はないのですが、`virtual`の後にスペースがないのは気持ち悪いです。

Sharoku-haga replied:
コードを修正しました。

masami-sugao commented on Twar/Twar/SubScene/Battle.h:
```diff
@@ -0,0 +1,45 @@
+	virtual void Draw();
+
+private:
+	GameDataManager* m_pGameDataManager;		//!< ゲームデータ管理クラス
```
`m_pGameDataManager`は「クラス」ではなく「GameDataManagerクラスのインスタンスへのポインタ」なのでこの変数に対するコメントとしてはこれは的確ではないです。

Sharoku-haga replied:
レビューありがとうございます。コメント修正しました。

masami-sugao reviewed (commented):
いきなりソースコードを見たので的外れな部分もあるかもしれません。
あくまで参考として見てもらえれば。

taishi-matsumura commented:
コンストラクタの初期化リストの件、[GoogleのStyleGuide](http://www.textdrop.net/google-styleguide-ja/cppguide.xml#%E3%82%B3%E3%83%B3%E3%82%B9%E3%83%88%E3%83%A9%E3%82%AF%E3%82%BF%E3%82%A4%E3%83%8B%E3%82%B7%E3%83%A3%E3%83%A9%E3%82%A4%E3%82%B6%E3%83%AA%E3%82%B9%E3%83%88)には
``` cpp
// 複数行にわたる場合、スペース4つでインデントして、イニシャライザの最初の行にコロンを入れる。
MyClass::MyClass(int var)
    : some_var_(var),             // スペース4つでインデント
      some_other_var_(var + 1) {  // 一列に並べる
  ...
  DoSomething();
  ...
}
```
のように書かれていますが、Visual Studioでこれを実現するのは難しそうなので、
```
    sample( const int a, const int b )
        : a_( a )
        , b_( b )
    { }
```
が良いように思います。(https://gist.github.com/mntone/7513279)
行頭のカンマはよくやるので慣れればOKかと。
